### PR TITLE
Search cert file and dir returned by openssl-probe

### LINKED ISSUE #28: FreeBSD certs not found
I have a FreeBSD 13 system and I have uploaded my custom certs into /usr/local/share/certs/ca-root-nss.crt which allows OOTB curl to work.  However, Rust programs (ex: rustup-init) built using rustls-native-certs do not check that location.  I can work around the issue by setting the env variable SSL_CERT_FILE=/usr/local/share/certs/ca-root-nss.crt but I would like for that location to be searched by this crate.

### PR DESCRIPTION
**Pull request by pgerber**
Fixes #28

### CODE REVIEW

**Review by djc — COMMENTED** [review]
Thanks for your work on improving rustls-native-certs!

Unfortunately in its current form this PR is quite hard to review. When we merge code we like make sure the PR consists of a series of well-defined changes, whereas this PR contains a bunch of changes all squashed together. Please read through https://github.com/rustls/rustls/blob/main/CONTRIBUTING.md for more guidance.

In particular, it would be good to avoid moving existing code around in the same commit as adding a bunch of new code, it would be good to split out addition of documentation for existing code from any other changes (in a separate commit). Then, further commits that change the behavior should update the documentation.

It would also be good to add a PR description that briefly describes which conceptual changes you're trying to make here.

**Comment by pgerber** [comment]
So, the change should now be in a more reviewable state. I wonder if I'm just too used to people squashing commits and using commit message like "fixes #11". I should remind myself more often to work with small, digestible commits, even found I few issues splitting things down into pieces.

File: src/lib.rs [hunk 1]
```diff
@@ -84,3 +126,97 @@ fn load_pem_certs(path: &Path) -> Result<Vec<CertificateDer<'static>>, Error> {
...
+    for entry in dir_reader {
+        let entry = entry?;
+        let path = entry.path();
+        let file_name = path
+            .file_name()
+            .expect("dir entry with no name");
+        let metadata = match fs::metadata(&path) {
+            Ok(metadata) => metadata,
+            Err(e) if e.kind() == ErrorKind::NotFound => {
+                // Dangling symlink
```
[the remark this answers is not in the document]
Comment (pgerber) [reply src/lib.rs:149]: Symlinks is how `openssl rehash works`.

Excerpt from manpage:

> openssl rehash scans directories and calculates a hash value of each .pem, .crt, .cer, or .crl file in
       the specified directory list and creates symbolic links for each file, where the name of the link is the
       hash value.  (If the platform does not support symbolic links, a copy is made.)  This command is useful
       as many programs that use OpenSSL require directories to be set up like this in order to find
       certificates

Not using symlinks seems like an exception. I'll add a comment.

File: tests/smoketests.rs [hunk 2]
```diff
@@ -108,3 +108,17 @@ fn badssl_with_env() {
...
+fn malformed_file_from_env() {
+    let site = "letsencrypt.org";
+    check_site(site).unwrap();
+
+    // Invalid data in file!() is ignored.
+    env::set_var("SSL_CERT_FILE", file!());
+    // Connect error (not cert load error).
+    assert!(check_site(site).is_err());
+
+    env::remove_var("SSL_CERT_FILE");
```
[the remark this answers is not in the document]
Comment (pgerber) [reply tests/smoketests.rs]: I'll have to think about this a bit more. Yes, `load_pem_certs()` should work for most cases, I guess. Of course, this would mean the tests could no longer be integration tests (`load_pem_certs()` is a private function). This is probably alright though.

I somehow do like the idea of having something close to integration tests. Perhaps, would could do something like this:

```
pub fn load_native_certs() -> Result<Vec<CertificateDer<'static>>, Error> {
    let ssl_cert_file = env::var("SSL_CERT_FILE").map(PathBuf::from);
    let ssl_cert_dir = ...;
    load_native_certs_internal(ssl_cert_file, ssl_cert_dir);
}
```

Then we could use `load_native_certs_internal()` (better suggestions for a name welcome) in the tests.

File: src/lib.rs (hunk 1 above)
[the remark this answers is not in the document]
Comment (pgerber) [reply src/lib.rs:149]: I guess technically, we could skip the `fs::metadata()` (which may do additional I/O) and just assume files matching the naming scheme resolve to regular files, and handle a not found error returned by `load_pem_certs()`. Not checking for not found seems dangerous, however, given that OpenSSL uses symlinks. Someone ought to remove a file without removing the symlink.

File: tests/smoketests.rs [hunk 3]
```diff
@@ -122,3 +122,29 @@ fn malformed_file_from_env() {
 
     env::remove_var("SSL_CERT_FILE");
 }
+
+#[test]
+#[serial]
+fn from_env_missing_file() {
+    env::set_var("SSL_CERT_FILE", Path::new("no/such/file"));
```
Comment (cpu) [inline tests/smoketests.rs:172]: Separate from the idea of a helper fn, I think L185 could use `no/such/directory` directly right? That would save the `Path` import being added on L3.

File: Cargo.toml [hunk 4]
```diff
@@ -19,6 +19,7 @@ ring = "0.17"
 rustls = "0.23"
```
Comment (cpu) [inline Cargo.toml:19]: note: small typo in your commit message for 83d9b88efa8d8ef3bfb18211765262ac3f5c718e:

"On Unix, certifictes were " -> "certifictes" replaced with "certificates"

File: src/lib.rs [hunk 5]
```diff
@@ -37,39 +37,81 @@ mod macos;
...
+use std::ffi::OsStr;
+use std::fs::{self, File};
 use std::io::BufReader;
 use std::io::{Error, ErrorKind};
 use std::path::{Path, PathBuf};
 
 use pki_types::CertificateDer;
 
+const ENV_CERT_FILE: &str = "SSL_CERT_FILE";
+const ENV_CERT_DIR: &str = "SSL_CERT_DIR";
```
Comment (cpu) [inline src/lib.rs:49]: For new code I think our [style guide](https://github.com/rustls/rustls/blob/main/CONTRIBUTING.md) wants these module private `const`'s at the bottom of the file.

File: src/lib.rs [hunk 6]
```diff
@@ -37,39 +37,81 @@ mod macos;
...
 /// Load root certificates found in the platform's native certificate store.
 ///
-/// If the SSL_CERT_FILE environment variable is set, certificates (in PEM
-/// format) are read from that file instead.
+/// ## Environment Variables
+///
+/// | Env. Var.      | Description                                                         |
+/// |----------------|---------------------------------------------------------------------|
+/// | SSL_CERT_FILE  | File containing an arbitrary number of certificates in PEM format.  |
+/// | SSL_CERT_DIR   | Directory utilizing the hierarchy used by OpenSSL's [c_rehash].     |
```
Comment (cpu) [inline src/lib.rs:58]: I might be tempted to say "hierarchy and naming convention used by ..." here to emphasize the file names are a wonky hash based doo-dad

File: src/lib.rs [hunk 7]
```diff
@@ -37,39 +37,81 @@ mod macos;
...
 ///
-/// If it is defined, it is always used, so it must be a path to an existing,
-/// accessible file from which certificates can be loaded successfully. However,
-/// [`rustls_pemfile::certs()`] will try to skip parts of a file which
-/// are not part of a certificate.
-fn load_certs_from_env() -> Option<Result<Vec<CertificateDer<'static>>, Error>> {
-    let cert_var_path = PathBuf::from(env::var_os(ENV_CERT_FILE)?);
+/// If SSL_CERT_DIR is defined, a directory must exist at this path, and all
+/// [hash files](`is_hash_file_name()`) contained in it must be loaded successfully,
+/// subject to the rules outlined above for SSH_CERT_FILE. The directory is not
```
Comment (cpu) [inline src/lib.rs:93]: small typo here: `SSH_CERT_FILE` -> `SSL_CERT_FILE`

File: src/lib.rs [hunk 8]
```diff
@@ -37,39 +37,81 @@ mod macos;
...
+/// file may contain no certificates.
 ///
-/// If it is defined, it is always used, so it must be a path to an existing,
-/// accessible file from which certificates can be loaded successfully. However,
-/// [`rustls_pemfile::certs()`] will try to skip parts of a file which
-/// are not part of a certificate.
-fn load_certs_from_env() -> Option<Result<Vec<CertificateDer<'static>>, Error>> {
-    let cert_var_path = PathBuf::from(env::var_os(ENV_CERT_FILE)?);
+/// If SSL_CERT_DIR is defined, a directory must exist at this path, and all
+/// [hash files](`is_hash_file_name()`) contained in it must be loaded successfully,
```
Comment (cpu) [inline src/lib.rs:92]: I think this link target doesn't work out when docs are built w/ `cargo doc --no-deps --document-private-items`, but this seems to work:

```
/// [hash files][is_hash_file_name] contained in it must be loaded successfully,
```

File: src/lib.rs [hunk 9]
```diff
@@ -37,39 +37,81 @@ mod macos;
...
+
+    let mut certs = Vec::new();
+    if let Some(file) = file {
+        certs.append(&mut load_pem_certs(&file)?);
+    }
+    if let Some(dir) = dir {
+        certs.append(&mut load_pem_certs_from_dir(&dir)?);
+    }
+
+    certs.sort_unstable_by(|a, b| a.cmp(b));
```
Comment (cpu) [inline src/lib.rs:111]: just double checking: does this do what we want given `CertificateDer` doesn't impl `Ord`? :thinking:

File: src/lib.rs [hunk 10]
```diff
@@ -37,39 +37,81 @@ mod macos;
...
+    certs.dedup_by(|a, b| a == b);
```
Comment (cpu) [inline src/lib.rs:112]: I think this is equivalent to `certs.dedup()` no?

File: src/lib.rs [hunk 11]
```diff
@@ -84,3 +126,97 @@ fn load_pem_certs(path: &Path) -> Result<Vec<CertificateDer<'static>>, Error> {
         })
         .collect()
 }
+
+/// Load certificate from certificate directory (what OpenSSL calls CAdir)
```
Comment (cpu) [inline src/lib.rs:130]: minor nit: I would be tempted to put "CAdir" in backticks like `CAdir`

File: src/lib.rs [hunk 12]
```diff
@@ -84,3 +126,97 @@ fn load_pem_certs(path: &Path) -> Result<Vec<CertificateDer<'static>>, Error> {
...
+///
+/// This directory can contain other files and directories. CAfile tends
```
Comment (cpu) [inline src/lib.rs:132]: Ditto re: backticks for `CAfile`.

File: src/lib.rs [hunk 13]
```diff
@@ -84,3 +126,97 @@ fn load_pem_certs(path: &Path) -> Result<Vec<CertificateDer<'static>>, Error> {
...
+/// is not verified, however.
+fn load_pem_certs_from_dir(dir: &Path) -> Result<Vec<CertificateDer<'static>>, Error> {
+    let dir_reader = fs::read_dir(dir)?;
+    let mut certs = Vec::new();
+    for entry in dir_reader {
+        let entry = entry?;
+        let path = entry.path();
+        let file_name = path
+            .file_name()
+            .expect("dir entry with no name");
```
Comment (cpu) [inline src/lib.rs:145]: I think it might be preferable to continue the iteration rather than panic in this (unlikely) case. Or, alternatively, since this is already a fn returning a `Result` we could cook up a `Error` to yield here.

File: src/lib.rs [hunk 14]
```diff
@@ -84,3 +126,97 @@ fn load_pem_certs(path: &Path) -> Result<Vec<CertificateDer<'static>>, Error> {
...
+        None => return false, // non-UTF8 can't be hex digits
+    };
+
+    if file_name.len() == 10 {
+        let mut iter = file_name.chars();
+        let iter = iter.by_ref();
+        iter.take(8)
+            .all(|c| c.is_ascii_hexdigit())
+            && iter.next() == Some('.')
+            && matches!(iter.next(), Some('0'..='9'))
```
Comment (cpu) [inline src/lib.rs:185]: optional nit: I think this last `matches!` might be a touch more readable using `is_ascii_digit` like:
```
&& matches!(iter.next(), Some(c) if c.is_ascii_digit())
```

File: src/unix.rs [hunk 15]
```diff
@@ -7,8 +7,17 @@ use std::io::Error;
...
+        Some(cert_file) => load_pem_certs(&cert_file)?,
+        None => Vec::new(),
+    };
+
+    if let Some(cert_dir) = likely_locations.cert_dir {
+        certs.append(&mut load_pem_certs_from_dir(&cert_dir)?);
     }
+
+    certs.sort_unstable_by(|a, b| a.cmp(b));
+    certs.dedup_by(|a, b| a == b);
```
Comment (cpu) [inline src/unix.rs:20]: same feedback here re `certs.dedup()` instead.

File: tests/smoketests.rs [hunk 16]
```diff
@@ -109,6 +111,39 @@ fn badssl_with_env() {
...
+    );
+    assert!(check_site("self-signed.badssl.com").is_err());
+
+    // OpenSSL uses symlinks too. So, use one for testing too, if possible.
+    #[cfg(unix)]
+    symlink(original, link1).unwrap();
+    #[cfg(not(unix))]
+    std::fs::copy(original, link1).unwrap();
+
+    // Dangaling symlink
```
Comment (cpu) [inline tests/smoketests.rs:149]: typo here: "Dangaling" -> "Dangling"

**Review by cpu — COMMENTED** [review]
Nice work! I had a handful of nits but I think the substance of the work is on the right track :+1:

File: src/lib.rs (hunk 1 above)
[the remark this answers is not in the document]
Comment (djc) [reply src/lib.rs:149]: I'm not saying we shouldn't resolve symlinks, just want to add a bit more context here through comments to make it clear that's an important part of what this code is trying to do.

File: src/lib.rs (hunk 13 above)
Comment (pgerber) [reply src/lib.rs:145]: I used `.expect()` here because I believe this to be unreachable. We are iterating over a directory and extracting `.file_name()`. As far as I know, there can be no file of any sort in a directory that doesn't have a name (except `"."` and `".."` which aren't yielded).

An alternative would be to use `entry.path()` which will yield an OsStr (not Option<OsStr>). It would mean an additional allocation but avoid the `.expect()`.

Let me know what you prefer.

Comment (cpu) [reply src/lib.rs:145]: That sounds reasonable, a comment with this context would be an OK way to resolve this feedback if you'd like to keep the opportunity expect 👍

File: src/lib.rs (hunk 9 above)
Comment (pgerber) [reply src/lib.rs:111]: When I tested it locally it worked. Here we rely on `Deref<Target = [u8]>`. I don't know if the DER representation can differ for the same certificate. (Can it be changed without invalidating it (e.g. by making the self-signature invalid))?

This code is here because it's likely that the same cert exists twice, once in CAfile and once in CAdir. Both are usually coming from the same source, the operating system. It seems unlikely that the two copies would differ.

I'll add a test. It will still be possible that we'll miss some exotic setup but having duplicate certificates won't be a big deal in most cases.

Comment (cpu) [reply src/lib.rs:111]: > Here we rely on Deref<Target = [u8]>

That's the main insight I was missing 👍 Thanks

File: src/lib.rs (hunk 12 above)
Comment (pgerber) [reply src/lib.rs:132]: I'm not sure. For me CAdir is more of a name used by OpenSSL. I usually use backticks only for some sort of code, either command line (`openssl reshash`), or a reference to the Rust code (`AStruct`, `func()`, `variable_name`).

File: src/lib.rs (hunk 8 above)
Comment (pgerber) [reply src/lib.rs:92]: When I wrote this, I wasn't sure about the syntax, so, I just ended up testing it and the link appeared to work just fine.

[Looking at the documentation](https://doc.rust-lang.org/rustdoc/write-documentation/linking-to-items-by-name.html#linking-to-items-by-name), there appear to be multiple options.

Comment (cpu) [reply src/lib.rs:92]: Apologies, it does appear to render correctly as-is when I restested just now.

File: src/lib.rs (hunk 12 above)
Comment (cpu) [reply src/lib.rs:132]: I don't feel strongly, leaving it as-is is OK with me :+1:

File: src/lib.rs [hunk 17]
```diff
@@ -62,8 +62,8 @@ const ENV_CERT_FILE: &str = "SSL_CERT_FILE";
```
[the remark this answers is not in the document]
Comment (pgerber) [reply src/lib.rs]: Done

File: Cargo.toml (hunk 4 above)
Comment (pgerber) [reply Cargo.toml:19]: Done

File: src/lib.rs (hunk 5 above)
Comment (pgerber) [reply src/lib.rs:49]: Done

File: src/lib.rs (hunk 6 above)
Comment (pgerber) [reply src/lib.rs:58]: Done

File: src/lib.rs (hunk 10 above)
Comment (pgerber) [reply src/lib.rs:112]: Done

File: src/lib.rs (hunk 13 above)
Comment (pgerber) [reply src/lib.rs:145]: Comment added

File: src/lib.rs (hunk 14 above)
Comment (pgerber) [reply src/lib.rs:185]: Done

File: src/unix.rs (hunk 15 above)
Comment (pgerber) [reply src/unix.rs:20]: Done

File: src/lib.rs (hunk 5 above)
Comment (djc) [reply src/lib.rs:49]: (Feel free to hit "Resolve conversation" if you're reasonably sure you've addressed the issue.)

File: tests/smoketests.rs (hunk 16 above)
Comment (pgerber) [reply tests/smoketests.rs:149]: Done

File: src/lib.rs [hunk 18]
```diff
@@ -84,3 +126,97 @@ fn load_pem_certs(path: &Path) -> Result<Vec<CertificateDer<'static>>, Error> {
         })
         .collect()
 }
+
+/// Load certificate from certificate directory (what OpenSSL calls CAdir)
+///
+/// This directory can contain other files and directories. CAfile tends
+/// to be in here too. To avoid loading something twice or something that
+/// isn't an valid certificate. Load is limited to those files that have
```
[the remark this answers is not in the document]
Comment (pgerber) [reply src/lib.rs:134]: Done

File: src/lib.rs (hunk 1 above)
[the remark this answers is not in the document]
Comment (pgerber) [reply src/lib.rs:149]: Added a comment.

File: src/unix.rs [hunk 19]
```diff
@@ -7,8 +7,17 @@ use std::io::Error;
 pub fn load_native_certs() -> Result<Vec<CertificateDer<'static>>, Error> {
     let likely_locations = openssl_probe::probe();
 
-    match likely_locations.cert_file {
-        Some(cert_file) => load_pem_certs(&cert_file),
-        None => Ok(Vec::new()),
+    let mut certs = match likely_locations.cert_file {
```
[the remark this answers is not in the document]
Comment (pgerber) [reply src/unix.rs:10]: Done

File: tests/smoketests.rs (hunk 2 above)
[the remark this answers is not in the document]
Comment (pgerber) [reply tests/smoketests.rs]: I have been thinking about this for a while, I believe `#[serial]` should be just fine. I realized, though, that many of the new tests just check certificate loading (e.g. ensure no certificate is found, or an error is returned). It seems rather silly, in hindsight, to use `check_site()` for those. I converted them into unit tests which turned out to be easier to read anyway.

File: tests/smoketests.rs (hunk 3 above)
[the remark this answers is not in the document]
Comment (pgerber) [reply tests/smoketests.rs:172]: Seems almost a bit overkill and we have to clear both vars at the beginning anyway, because Cargo may have set them. I'll leave things as they are.

File: src/lib.rs (hunk 17 above)
Comment (cpu) [inline src/lib.rs]: Small typo in the commit message, "file must exists but mey be a non-regular" - "mey" :arrow_right:  "may"

File: tests/smoketests.rs (hunk 3 above)
Comment (cpu) [reply tests/smoketests.rs:172]: I still see `Path` being imported and used on L129 of d4ffe71475cb748703d14ed00d1e646d28ced696 Did that change get dropped?

File: Cargo.toml [hunk 20]
```diff
@@ -19,6 +19,7 @@ ring = "0.17"
...
 rustls-webpki = "0.102"
 serial_test = "3"
+tempfile = "3.5"
```
Comment (cpu) [inline Cargo.toml:22]: Would you mind folding the `Cargo.lock` update into this commit? I think they should be done in sync so that this commit builds cleanly with `--locked`

File: src/lib.rs (hunk 5 above)
Comment (cpu) [reply src/lib.rs:49]: I still see these `consts` at the top of the file. Maybe forgot a push?

**Review by cpu — COMMENTED** [review]
I started reviewing the branch, but it seems like most of the resolved feedback isn't present. Is it possible you pushed the wrong branch? 

I will pick up my re-review once that's sorted. Thanks!

**Comment by pgerber** [comment]
@cpu, updated commits will follow soon. I'm afraid I had to leave before I managed to do a final review before pushing the changes.

**Review by cpu — APPROVED** [review]
Thanks, looks good to me :rocket:

File: tests/smoketests.rs [hunk 21]
```diff
@@ -8,7 +8,15 @@ use std::{env, panic};
 // the global env var configuration in the env var test interferes with the others.
 use serial_test::serial;
 
-fn check_site(domain: &str) {
+/// Check if connection to site works
+///
+/// Tries to returns `Err` if, and only if, there is an issue
+/// connecting due to a certificate issue.
```
Comment (djc) [inline tests/smoketests.rs:18]: "Tries to returns `Err`" doesn't read quite right. Maybe:

> Yields an `Err` if and only if there is an issue connecting that
> appears to be due to a certificate problem.

File: src/lib.rs [hunk 22]
```diff
@@ -63,7 +63,9 @@ const ENV_CERT_FILE: &str = "SSL_CERT_FILE";
 /// Returns None if SSL_CERT_FILE is not defined in the current environment.
 ///
 /// If it is defined, it is always used, so it must be a path to a real
-/// file from which certificates can be loaded successfully.
+/// file from which certificates can be loaded successfully. However,
```
Comment (djc) [inline src/lib.rs]: I'd skip the "However" here, which doesn't seem entirely appropriate? Maybe rephrase as:

> While parsing, [`rustls_pemfile::certs()`] parser will ignore parts of the file which are not part of a certificate.

File: Cargo.lock [hunk 23]
```diff
@@ -52,15 +52,15 @@ dependencies = [
 
 [[package]]
 name = "autocfg"
-version = "1.1.0"
+version = "1.3.0"
```
Comment (djc) [inline Cargo.lock:55]: This seems to have a bunch of unrelated `Cargo.lock` changes, can we minimize this to what is necessary for tempfile?

File: src/lib.rs [hunk 24]
```diff
@@ -46,30 +47,80 @@ use pki_types::CertificateDer;
...
+        certs.append(&mut load_pem_certs_from_dir(cert_dir)?);
+    }
+
+    certs.sort_unstable_by(|a, b| a.cmp(b));
+    certs.dedup();
+
+    Ok(certs)
+}
+
+struct CertPaths {
```
Comment (djc) [inline src/lib.rs:121]: I'm not sure this type is useful? We can just pass two arguments to `load_certs_from_paths()`?

File: src/lib.rs [hunk 25]
```diff
@@ -46,30 +47,80 @@ use pki_types::CertificateDer;
...
+        file: env::var_os(ENV_CERT_FILE).map(PathBuf::from),
+        dir: env::var_os(ENV_CERT_DIR).map(PathBuf::from),
+    };
+
+    if paths.file.is_none() && paths.dir.is_none() {
+        return Ok(None);
+    }
 
-    Some(load_pem_certs(&cert_var_path))
+    Ok(Some(load_certs_from_paths(&paths)?))
```
Comment (djc) [inline src/lib.rs:102]: Suggest simplifying this to:

```rust
let file = env::var_os(ENV_CERT_FILE).map(PathBuf::from);
let dir = env::var_os(ENV_CERT_DIR).map(PathBuf::from);
Ok(match (&file, &dir) {
     (None, None) => None,
     _ => Some(load_certs_from_paths(file, dir)?),
})
```

File: tests/compare_mozilla.rs [hunk 26]
```diff
@@ -140,6 +142,8 @@ fn test_contains_most_roots_known_by_mozilla() {
 
 #[test]
 fn util_list_certs() {
+    common::clear_env();
```
Comment (djc) [inline tests/compare_mozilla.rs:145]: This test is not annotated with `#[serial]` but it does use `clear_env()` which AIUI is undefined behavior:

> In multi-threaded programs, you must ensure that are no other threads concurrently writing or reading(!) from the environment through functions other than the ones in this module. You are responsible for figuring out how to achieve this, but we strongly suggest not using set_var or remove_var in multi-threaded programs at all.

We should probably document this on `clear_env()`?

**Review by djc — APPROVED** [review]
Looking good, some hopefully minor feedback left.

File: tests/compare_mozilla.rs (hunk 26 above)
Comment (pgerber) [reply tests/compare_mozilla.rs:145]: Will do.

I guess that's for another time but I wonder if I should look into adding a `serial_test::is_serial()` for checking if we are in a `#[serial]` context. Then, we could just add an assert in `clear_env()`.

File: src/lib.rs (hunk 24 above)
Comment (pgerber) [reply src/lib.rs:121]: The reason I added the struct is that both arguments are of the same type. Calling `load_certs_from_paths(file, dir)` instead of `load_certs_from_paths(dir, file)` will compile and is hard to spot in reviews.

Want me to remove the type?

File: src/lib.rs (hunk 22 above)
Comment (pgerber) [reply src/lib.rs]: "However" I chose because I'm trying to say certificate loading needs to be successful but the fact that parts of the file may be ignored (e.g. because the wrong certificate format is used or a certificate is corrupted) can mean that we report success/`Ok(_)` when really the certificate failed to parse, at least from a user's perspective.

What about this:

> While parsing, [rustls_pemfile::certs()] parser will ignore parts of the file which are not considered part of a certificate. Certificates which are not in the right format (PEM) or are otherwise corrupted may get ignored silently.

I also wonder if this should we document this in the public part of the documentation too. Perhaps we could clarify what format is expected to make it less likely people will use the wrong format. Something like:

Be sure certificates are in PEM format:

```
<insert example cert in PEM format>
```

Comment (pgerber) [reply src/lib.rs]: What's an error and what's ignored seems a bit arbitrary to me.

This is ignored

```
----BEGIN CERTIFICATE-----
^ missing '-' at start
<content>
-----END CERTIFICATE-----
```

This is an error:

```
-----BEGIN CERTIFICATE-----
<content>
-----END CERTIFICATE----
                        ^ missing '-'
```

This is ignored:

```
-----BEGIN CERTIFICATE-----
<content>
-----END CERTIFICATE----
                        ^ missing '-' AND '\n'
```

This is ignore:

```
-----BEGIN CERTIFICATE-----
<content>
-----END CERTIFICATE-----
                         ^ missing '\n'
```

File: src/lib.rs (hunk 24 above)
Comment (djc) [reply src/lib.rs:121]: Hmm, seems like a good rationale to keep it -- sorry for the noise!

File: src/lib.rs [hunk 27]
```diff
@@ -46,30 +47,80 @@ use pki_types::CertificateDer;
...
+}
+
+fn load_certs_from_paths(paths: &CertPaths) -> Result<Vec<CertificateDer<'static>>, Error> {
```
Comment (djc) [inline src/lib.rs:105]: If we're going to keep the `CertPaths` type, maybe turn this into `CertPaths::load()`?

File: src/lib.rs (hunk 22 above)
Comment (djc) [reply src/lib.rs]: I think we'd probably take PRs to make all of those errors, but the main goal in that crate was to properly ignore things that don't look like PEM objects at all which has been sighted in the wild. I guess these cases really haven't shown up.

All of your documentation suggestions sound good to me!

File: tests/compare_mozilla.rs (hunk 26 above)
Comment (pgerber) [reply tests/compare_mozilla.rs:145]: I ended up marking clear_env() unsafe. It's up to the caller to ensure safety, so it really should be marked this way, I think.

File: Cargo.lock (hunk 23 above)
Comment (pgerber) [reply Cargo.lock]: Okay, changes are now limited to the necessary changes only. If I remember right, I updated everything after the Winows or Mac pipeline failed to compile. Let's hope it works now. Those OSes are not exactly readability available and cross-compiling is still one of Rust's weaknesses.

File: src/lib.rs (hunk 22 above)
Comment (pgerber) [reply src/lib.rs]: (Still pending)

File: Cargo.lock (hunk 23 above)
Comment (djc) [reply Cargo.lock]: Seems like CI is happy!

File: src/lib.rs (hunk 22 above)
Comment (pgerber) [reply src/lib.rs]: PEM appears to use the encoding defined in [RFC7468](https://datatracker.ietf.org/doc/html/rfc7468). Ignoring invalid sections of the file is part of the encoding but, if I read the grammar right, the line feed at the end of `-----END CERTIFICATE-----` is explicitly optional. So, it should probably be made optional in RusTLS parser too.

Comment (djc) [reply src/lib.rs]: Thanks for looking into that! Would be good to file a followup issue and/or PR for that.

**Comment by djc** [comment]
Thanks for all your work on this! Will do a follow-up which bumps the version and has some very minor tweaks.

File: src/lib.rs (hunk 22 above)
Comment (pgerber) [reply src/lib.rs]: Added it to my todo list.

**Comment by pgerber** [comment]
@djc, @cpu, thanks for making this pull request a pleasant experience.

**Comment by cpu** [comment]
My pleasure :-) Thanks again for the contribution!
